### PR TITLE
[MNOE-872] Session timeout modal

### PR DESCRIPTION
**Pull request by enizor**
Relies on [MNOE #632](https://github.com/maestrano/mno-enterprise/pull/632)

### CODE REVIEW

File: src/app/components/mno-session-timeout/mno-session-timeout.html [hunk 1]
```diff
@@ -0,0 +1,14 @@
+<div class="modal-header">
+  <button type="button" class="close" ng-click="logOff()"></button>
+  <h4 class="modal-title" translate >{{'mnoe_admin_panel.auth.sessions.timeout.header'}}</h4>
+</div>
+<div class="modal-body">
+  <p translate>{{countdown == 0? 'mnoe_admin_panel.auth.sessions.timeout.body.must_logout' : 'mnoe_admin_panel.auth.sessions.timeout.body.can_login'}}</p>
```
Comment (ouranos) [inline src/app/components/mno-session-timeout/mno-session-timeout.html]: `countdown == 0 ? '...`

File: src/app/components/mno-session-timeout/mno-session-timeout.svc.coffee [hunk 2]
```diff
@@ -0,0 +1,50 @@
+@App.service 'MnoSessionTimeout', ($q, $timeout, $uibModal, DEVISE_CONFIG) ->
+  _self = @
+
+  timer = null
+
+  @resetTimer = ->
+    _self.cancelTimer()
+    console.debug 'reset timer'
+    timer = $timeout(showTimeoutModal, DEVISE_CONFIG.timeout_in * 1000 - 12 * 1000)
```
Comment (ouranos) [inline src/app/components/mno-session-timeout/mno-session-timeout.svc.coffee]: `(DEVISE_CONFIG.timeout_in - 12) * 1000`

File: src/locales/en-AU.json [hunk 3]
```diff
@@ -333,6 +333,13 @@
   "mnoe_admin_panel.dashboard.mno_confirm.ok": "OK",
   "mnoe_admin_panel.dashboard.mno_confirm.proceed": "Proceed?",
   "mnoe_admin_panel.dashboard.mno_confirm.perform": "Perform this action?",
+  "mnoe_admin_panel.auth.sessions.timeout.header": "Session Timeout",
+  "mnoe_admin_panel.auth.sessions.timeout.body.can_login": "You are being logged off due to inactivity. Please choose to stay signed in or to logoff. Otherwise, you will be logged off automatically",
+  "mnoe_admin_panel.auth.sessions.timeout.body.must_logout": "You are logged off due to inactivity. Please log in again by using the button below or by refreshing the page",
+  "mnoe_admin_panel.auth.sessions.timeout.stay_logged_in": "Stay Logged In",
+  "mnoe_admin_panel.auth.sessions.timeout.log_off": "Log Off",
+  "mnoe_admin_panel.auth.sessions.timeout.log_in": "Log In",
+  "mnoe_admin_panel.auth.sessions.timeout.error": "We couldnt manage to remain you logged in. You will be redirected to the login page",
```
Comment (ouranos) [inline src/locales/en-AU.json]: "couldn't" or "could not"

File: src/app/index.config.coffee [hunk 4]
```diff
@@ -52,6 +52,18 @@
     )
   )
 
+  .config ($httpProvider) ->
+    $httpProvider.interceptors.push ($q, $injector, DEVISE_CONFIG) ->
+      {
+        request: (config) ->
+          # Intercept requests made to the API and reset the sessions timeout
+          if DEVISE_CONFIG.timeout_in > 0
+            if config.url.includes('mnoe') && !config.url.includes('sign_out') && !config.url.includes('sign_in')
```
Comment (ouranos) [inline src/app/index.config.coffee]: Is that more readable:
```coffeescript
if config.url.includes('mnoe') && !config.url.match(/(sign_out|sign_in)/)
```
Also shouldn't the timer be reset at `sign_in`?

Comment (enizor) [reply src/app/index.config.coffee]: The timer does not need to be reseted on GET `/sign_in`. After a POST `sign_in`, the organization will be requested and reset the timer.
The regex isn't complex, I will use it

File: src/app/components/mno-session-timeout/mno-session-timeout.html [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+</div>
+<div class="modal-footer">
+  <button type="button" class="btn btn-default" ng-click="logOff()">{{ countdown == 0? 'mnoe_admin_panel.auth.sessions.timeout.log_in' : 'mnoe_admin_panel.auth.sessions.timeout.log_off'| translate}}</button>
+  <button type="button" class="btn btn-primary" ng-click="stayLoggedIn()" ng-disabled="isLoading" ng-show="countdown != 0">
```
Comment (manu-d) [inline src/app/components/mno-session-timeout/mno-session-timeout.html]: Same here

File: src/app/components/mno-session-timeout/mno-session-timeout.svc.coffee [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+  showTimeoutModal = ->
+    console.debug("Timeout finished!")
+
+    $uibModal.open({
+      size: 'md'
+      keyboard: false
+      backdrop: 'static'
+      templateUrl: 'app/components/mno-session-timeout/mno-session-timeout.html'
+      controller: modalController
+      })
```
Comment (manu-d) [inline src/app/components/mno-session-timeout/mno-session-timeout.svc.coffee]: Indentation

File: src/app/components/mno-session-timeout/mno-session-timeout.svc.coffee [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+    $scope.stayLoggedIn = () ->
+      $scope.isLoading = true
+      MnoeCurrentUser.refreshUser().then(
+        (response) ->
+          $uibModalInstance.close(response)
+        (error) ->
+          toastr.warning("mno_enterprise.auth.sessions.timeout.error")
+          $uibModalInstance.dismiss('cancel')
+          $http.delete('/mnoe/auth/users/sign_out')
+          $http.get('/mnoe/auth/users/sign_in')
```
Comment (manu-d) [inline src/app/components/mno-session-timeout/mno-session-timeout.svc.coffee]: You can call $scope.logOff() instead of the three previous lines

**Comment by MAhsenArif** [comment]
This can be closed as task is continued on https://github.com/maestrano/mnoe-admin-panel/pull/220

**Pull request closed without merging**